src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java

```diff
@@ -34,12 +38,11 @@
 public class AvailableGames {
   private static final boolean delayedParsing = false;
   private static final String ZIP_EXTENSION = ".zip";
-  private final TreeMap<String, URI> availableGames = new TreeMap<>();
-  private final Set<String> availableMapFolderOrZipNames = new HashSet<>();
+  private final Map<String, URI> availableGames = Collections.synchronizedMap(new TreeMap<>());
+  private final Set<String> availableMapFolderOrZipNames = Collections.synchronizedSet(new HashSet<>());
```

Comment:
It seems that once an `AvailableGames` instance is created, it is never again accessed concurrently.  Therefore, it shouldn't be necessary to use thread-safe collections for the entire lifetime of an `AvailableGames` instance.  They only need to be thread safe while `populateAvailableGames()` is running.

My suggestion would be to separate the initialization of these collections from the `AvailableGames` instance by using a factory method:

```diff
diff --git a/src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java b/src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java
index 4b53f1552..e6dd6fa42 100644
--- a/src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java
+++ b/src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java
@@ -38,11 +38,20 @@ import games.strategy.util.UrlStreams;
 public class AvailableGames {
   private static final boolean delayedParsing = false;
   private static final String ZIP_EXTENSION = ".zip";
-  private final Map<String, URI> availableGames = Collections.synchronizedMap(new TreeMap<>());
-  private final Set<String> availableMapFolderOrZipNames = Collections.synchronizedSet(new HashSet<>());
+  private final Map<String, URI> availableGames;
+  private final Set<String> availableMapFolderOrZipNames;
 
-  AvailableGames() {
-    populateAvailableGames(availableGames, availableMapFolderOrZipNames, Collections.synchronizedSet(new HashSet<>()));
+  private AvailableGames(final Map<String, URI> availableGames, final Set<String> availableMapFolderOrZipNames) {
+    this.availableGames = new TreeMap<>(availableGames);
+    this.availableMapFolderOrZipNames = new HashSet<>(availableMapFolderOrZipNames);
+  }
+
+  static AvailableGames newInstance() {
+    final Map<String, URI> availableGames = Collections.synchronizedMap(new TreeMap<>());
+    final Set<String> availableMapFolderOrZipNames = Collections.synchronizedSet(new HashSet<>());
+    final Set<String> mapNamePropertyList = Collections.synchronizedSet(new HashSet<>());
+    populateAvailableGames(availableGames, availableMapFolderOrZipNames, mapNamePropertyList);
+    return new AvailableGames(availableGames, availableMapFolderOrZipNames);
   }
 
   List<String> getGameNames() {
diff --git a/src/main/java/games/strategy/engine/framework/headlessGameServer/HeadlessGameServer.java b/src/main/java/games/strategy/engine/framework/headlessGameServer/HeadlessGameServer.java
index 7a08d05b9..9cff1f4b8 100644
--- a/src/main/java/games/strategy/engine/framework/headlessGameServer/HeadlessGameServer.java
+++ b/src/main/java/games/strategy/engine/framework/headlessGameServer/HeadlessGameServer.java
@@ -393,7 +393,7 @@ public class HeadlessGameServer {
       logger.info("Running ShutdownHook.");
       shutdown();
     }));
-    m_availableGames = new AvailableGames();
+    m_availableGames = AvailableGames.newInstance();
     m_gameSelectorModel = new GameSelectorModel();
     final String fileName = System.getProperty(GameRunner.TRIPLEA_GAME_PROPERTY, "");
     if (fileName.length() > 0) {
```

You could avoid the defensive copies in the constructor by refactoring a bit more to make the `populateFromDirectory()` and `populateFromFile()` methods to accept a `lock` parameter, but the above is probably the simplest thing that will work.

Reply:
I actually found an even easier solution...
See https://github.com/triplea-game/triplea/pull/2571/commits/d047201f9ee4e6138282944d0b24e12b7b988115

Reply:
Yes, that was what I was thinking about.  However, it wasn't clear to me that it is 100% thread safe because there doesn't appear to be a memory barrier to ensure the thread executing `AvailableGames#AvailableGames()` sees the effects of the background threads.  The act of copying a synchronized collection would trigger such a memory barrier and guarantee the constructor thread deterministically sees the final state of those collections.

It appears that `Future#get()` (which is invoked for all background threads via `invokeAll()`) has a _happens-before_ guarantee, which implies it effectively sets up a memory barrier between all background threads and the constructor thread.  So, in its current state, this _should_ be thread safe.

My only remaining suggestion in this area would be to encapsulate all the concurrency-related code in `populateAvailableGames()` so it's localized in one place, and a code reader doesn't have to back up 50 lines to see that these collections are thread-safe.  Also, the third parameter is for internal use by the method and can be removed:

```diff
diff --git a/src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java b/src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java
index 5ea70b285..ca959f1eb 100644
--- a/src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java
+++ b/src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java
@@ -41,10 +41,7 @@ public class AvailableGames {
   private final Set<String> availableMapFolderOrZipNames = new HashSet<>();
 
   AvailableGames() {
-    populateAvailableGames(
-        Collections.synchronizedMap(availableGames),
-        Collections.synchronizedSet(availableMapFolderOrZipNames),
-        Collections.synchronizedSet(new HashSet<>()));
+    populateAvailableGames(availableGames, availableMapFolderOrZipNames);
   }
 
   Set<String> getGameNames() {
@@ -77,8 +74,11 @@ public class AvailableGames {
     return Optional.ofNullable(availableGames.get(gameName)).map(Object::toString).orElse(null);
   }
 
-  private static void populateAvailableGames(final Map<String, URI> availableGames,
-      final Set<String> availableMapFolderOrZipNames, final Set<String> mapNamePropertyList) {
+  private static void populateAvailableGames(final Map<String, URI> outAvailableGames,
+      final Set<String> outAvailableMapFolderOrZipNames) {
+    final Map<String, URI> availableGames = Collections.synchronizedMap(outAvailableGames);
+    final Set<String> availableMapFolderOrZipNames = Collections.synchronizedSet(outAvailableMapFolderOrZipNames);
+    final Set<String> mapNamePropertyList = Collections.synchronizedSet(new HashSet<>());
     System.out.println("Parsing all available games (this could take a while). ");
     final List<File> files = allMapFiles();
     if (files.size() > 0) {
```

There's probably better choices for the parameter names.